tracer/src/Datadog.Trace/Debugger/Expressions/ProbeExpressionEvaluator.cs

```diff
@@ -0,0 +1,233 @@
+// <copyright file="ProbeExpressionEvaluator.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Linq.Expressions;
+using Datadog.Trace.Debugger.Configurations.Models;
+using Datadog.Trace.Debugger.Models;
+using Datadog.Trace.Util;
+
+namespace Datadog.Trace.Debugger.Expressions;
+
+internal class ProbeExpressionEvaluator
+{
+    internal ProbeExpressionEvaluator(
+        DebuggerExpression[] templates,
+        DebuggerExpression? condition,
+        DebuggerExpression? metric,
+        MethodScopeMembers scopeMembers)
+    {
+        Templates = templates;
+        Condition = condition;
+        Metric = metric;
+        ScopeMembers = scopeMembers;
+        CompiledTemplates = new Lazy<CompiledExpression<string>[]>(CompileTemplates, true);
+        CompiledCondition = new Lazy<CompiledExpression<bool>?>(CompileCondition, true);
+        CompiledMetric = new Lazy<CompiledExpression<double>?>(CompileMetric, true);
+    }
+
+    internal MethodScopeMembers ScopeMembers { get; }
+
+    internal DebuggerExpression[] Templates { get; }
+
+    internal DebuggerExpression? Condition { get; }
+
+    internal DebuggerExpression? Metric { get; }
+
+    internal Lazy<CompiledExpression<string>[]> CompiledTemplates { get; set; }
+
+    internal Lazy<CompiledExpression<bool>?> CompiledCondition { get; set; }
+
+    internal Lazy<CompiledExpression<double>?> CompiledMetric { get; set; }
+
+    internal ExpressionEvaluationResult Evaluate()
+    {
+        ExpressionEvaluationResult result = default;
+        EvaluateTemplates(ref result);
+        EvaluateCondition(ref result);
+        EvaluateMetric(ref result);
+        return result;
+    }
+
+    private void EvaluateTemplates(ref ExpressionEvaluationResult result)
+    {
+        var resultBuilder = StringBuilderCache.Acquire(StringBuilderCache.MaxBuilderSize);
```

Comment:
If an exception will be thrown in the following try/catch block, `resultBuilder` won't be released. I'd suggest adding a `finally` block that will release `resultBuilder`.